pipeline-runner/tests/testthat/test-gem2s-2-load_cellranger.R

```diff
@@ -20,7 +47,164 @@ test_that("format_annot deduplicates name column", {
+                           symbol = symbols)
+
+    outdir <- tempdir()
+    sample <- 'sample_a'
+    sample_dir <- file.path(outdir, sample)
+    dir.create(sample_dir)
+
+    mock_cellranger_files(counts, features, sample_dir)
+
+    prev_out <- list(config = list(samples = sample))
+    annot <- load_cellranger_files(NULL, NULL, prev_out, outdir)$output$annot
+
+    # unique gene names is same as number of gene names
+    expect_length(unique(annot$name), length(symbols))
+
+    # unique original names is same as unique gene names
+    expect_length(unique(annot$original_name), length(unique(symbols)))
+    unlink(sample_dir, recursive = TRUE)
+})
+
+test_that("load_cellranger uses appropriate feature columns", {
+
+    counts <- mock_counts()
+
+    symbols <- row.names(counts)
+    features <- data.frame(ensid = paste0('ENSFAKE', seq_len(nrow(counts))),
+                           symbol = symbols)
+
+    outdir <- tempdir()
+    sample <- 'sample_a'
+    sample_dir <- file.path(outdir, sample)
+    dir.create(sample_dir)
+
+    mock_cellranger_files(counts, features, sample_dir)
+
+    prev_out <- list(config = list(samples = sample))
+    out <- load_cellranger_files(NULL, NULL, prev_out, outdir)$output
+
+    # ensembl ids are counts row names
+    expect_equal(
+        features$ensid,
+        row.names(out$counts_list[[1]])
+    )
+
+    # ensembl ids are in column 'input' of annot
+    expect_equal(out$annot$input, features$ensid)
+
+    # symbols are in column 'name' of annot
+    expect_equal(out$annot$name, symbols)
+
+    unlink(sample_dir, recursive = TRUE)
+})
+
+
+test_that("load_cellranger loads multisample experiments", {
+
+    counts <- mock_counts()
+
+    symbols <- row.names(counts)
+    ensids <- paste0('ENSFAKE', seq_len(nrow(counts)))
+
+    features <- data.frame(ensid = ensids, symbol = symbols)
+
+    # most overlapping, some unique
+    ensids2 <- ensids
+    symbols2 <- symbols
+    ensids2[1:20] <- paste0(ensids2[1:20], '_2')
+    symbols2[1:20] <- paste0(symbols2[1:20], '_2')
+
+    features2 <- data.frame(ensid = ensids2, symbols = symbols2)
+
+    outdir <- tempdir()
+    samples <- c('sample_a', 'samble_b')
+    sample_dirs <- file.path(outdir, samples)
+    dir.create(sample_dirs[1])
+    dir.create(sample_dirs[2])
+
+    mock_cellranger_files(counts, features, sample_dirs[1])
+    mock_cellranger_files(counts, features2, sample_dirs[2])
+
+    prev_out <- list(config = list(samples = samples))
+    out <- load_cellranger_files(NULL, NULL, prev_out, outdir)$output
+
+    # loaded both
+    expect_equal(names(out$counts_list), samples)
+
+    # kept only unique ensembl ids
+    unique_ensids <- unique(c(ensids, ensids2))
+    expect_length(out$annot$input, length(unique_ensids))
+
+    # removed duplicated ensembl ids
+    expect_lt(
+        length(unique_ensids),
+        length(c(ensids, ensids2))
+    )
+
+    unlink(sample_dirs, recursive = TRUE)
+})
```

Comment:
How about checking for duplicated ens Ids parsing? We've had some bugs where this was the case

Reply:
Good suggestion - added.